src/stream/src/executor/over_window/frame_finder.rs

```diff
@@ -0,0 +1,893 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Helper functions for finding affected ranges from over window range cache for a
+//! given set of changes (delta).
+
+use std::ops::Bound;
+
+use delta_btree_map::DeltaBTreeMap;
+use risingwave_common::row::OwnedRow;
+use risingwave_expr::window_function::{FrameBound, RowsFrameBounds};
+
+use super::over_partition::CacheKey;
+
+// -------------------------- ↓ PUBLIC INTERFACE ↓ --------------------------
+
+/// Merge several `ROWS` frames into one super frame. The returned super frame is
+/// guaranteed to be *canonical*, which means that the `CURRENT ROW` is always
+/// included in the returned frame.
+pub(super) fn merge_rows_frames(rows_frames: &[&RowsFrameBounds]) -> RowsFrameBounds {
+    if rows_frames.is_empty() {
+        // When there's no `ROWS` frame, for simplicity, we don't return `None`. Instead,
+        // we return `ROWS BETWEEN CURRENT ROW AND CURRENT ROW`, which is the implicit
+        // frame for all selected upstream columns.
+        //
+        // For example, the following two queries are equivalent:
+        //
+        // ```sql
+        // SELECT a, b, sum(c) OVER (...) FROM t;
+        // SELECT
+        //   first_value(a) OVER (... ROWS BETWEEN CURRENT ROW AND CURRENT ROW),
+        //   first_value(b) OVER (... ROWS BETWEEN CURRENT ROW AND CURRENT ROW),
+        //   sum(c) OVER (...)
+        // FROM t;
+        // ```
+        //
+        // See https://risingwave.com/blog/risingwave-window-functions-the-art-of-sliding-and-the-aesthetics-of-symmetry/
+        // for more details.
+        return RowsFrameBounds {
+            start: FrameBound::CurrentRow,
+            end: FrameBound::CurrentRow,
+        };
+    }
+
+    let none_as_max_cmp = |x: &Option<usize>, y: &Option<usize>| match (x, y) {
+        // None means unbounded, which should be the largest.
```

Comment:
Did we consider the difference of semantics for query with/without `order by`? That is, the upper bound is

- `unbounded` by default for non-ordered windows, as assumed in this code
- `current row` by default for ordered windows

> https://www.postgresql.org/docs/current/tutorial-window.html
> 
> Above, since there is no `ORDER BY` in the `OVER` clause, the window frame is the same as the partition, which for lack of `PARTITION BY` is the whole table; in other words each sum is taken over the whole table and so we get the same result for each output row. But if we add an `ORDER BY` clause, we get very different results:
> 
> ```
> SELECT salary, sum(salary) OVER (ORDER BY salary) FROM empsalary;
>  salary |  sum
> --------+-------
>    3500 |  3500
>    3900 |  7400
>    4200 | 11600
>    4500 | 16100
>    4800 | 25700
>    4800 | 25700
>    5000 | 30700
>    5200 | 41100
>    5200 | 41100
>    6000 | 47100
> (10 rows)
> ```

Reply:
Yes, we use different default frames for queries with/without `ORDER BY`.

But we behave slightly different from PG, in that default frame in PG is `RANGE` frame while in RW is `ROWS` frame. This is mentioned in our document and some issues: https://github.com/risingwavelabs/risingwave/issues/11593 https://github.com/risingwavelabs/risingwave/issues/12822. We will adjust to align with PG after `RANGE` frame support merged.